### PR TITLE
The number of IR nodes is configurable

### LINKED ISSUE #73: Number of CNs and IR nodes should be configurable
In different development and testing scenarios we require different number of IR/CN/other nodes. It should be configurable and different setup scripts should be used depending on those settings.

### PR DESCRIPTION
**Pull request by End-rey**
Closes #73.

### CODE REVIEW

**Comment by End-rey** [comment]
@cthulhu-rider Now it shows that it is running, although there are a lot of errors in the ir nodes. Can you check the configs and see if I'm doing everything right?

**Comment by End-rey** [comment]
@cthulhu-rider Everything works. However, after the launch, the following errors appear in the `ir` node:
```
debug    network/server.go:1216  p2p notary request      {"error": "already exists in mempool", "hash": "22c5d8035255adf13142eec3e1f1f3256fa8d6874d1806738af44f0601939b09", "main": "dec54cf2feecb63c96ee8ce7f1c372b13bf626e15a0f1acb8dc0991e2811e42d"}

error  governance/process_update.go:82  can't update inner ring list with new alphabet keys  {"error": "test invocation failed: unsupported parameter noderoles.Role"}

error  governance/process_update.go:91  can't update list of notary nodes in FS chain  {"error": "test invocation failed: unsupported parameter noderoles.Role"}

error  governance/process_update.go:100  can't update list of alphabet nodes in neofs contract  {"error": "could not invoke alphabetUpdate: chain/client: contract execution finished with state FAULT; exception: at instruction 3666 (THROW): unhandled exception: \"this method must be invoked by alphabet\""}

error  governance/process_update.go:100  can't update list of alphabet nodes in neofs contract  {"error": "could not invoke alphabetUpdate: chain/client: contract execution finished with state FAULT; exception: at instruction 312 (SYSCALL): System.Contract.Call failed: called contract 6a715fe77786ad16a5f543e45c7ba242f196d01b not found: key not found"}

```
and in `storage` node:
```
debug      calculator/calls.go:231 all daughters trust iterator's init failure     {"error": "no data in 1 epoch for daughters"}
debug   calculator/calls.go:62  consumers trust iterator's init failure {"epoch": 1, "iteration": 1, "error": "no data for 0 iteration in 1 epoch for consumers's trusts"}
debug   calculator/calls.go:120 daughter trust iterator's init failure  {"error": "no data in 2 epoch for daughter: o5H2uGUa7cwBvb1GSx62SUL8Mkpp4PbgB1oUstPgnfo8"}
debug   calculator/calls.go:62  consumers trust iterator's init failure {"epoch": 1, "iteration": 2, "error": "no data for 1 iteration in 1 epoch for consumers's trusts"}
debug   calculator/calls.go:62  consumers trust iterator's init failure {"epoch": 2, "iteration": 2, "error": "no data for 1 iteration in 2 epoch for consumers's trusts"}
debug   calculator/calls.go:62  consumers trust iterator's init failure {"epoch": 1, "iteration": 3, "error": "no data for 2 iteration in 1 epoch for consumers's trusts"}
debug   calculator/calls.go:62  consumers trust iterator's init failure {"epoch": 2, "iteration": 3, "error": "no data for 2 iteration in 2 epoch for consumers's trusts"}
debug   calculator/calls.go:120 daughter trust iterator's init failure  {"error": "no data in 3 epoch for daughter: 2Bsv2nitVMDi9S77fuYNqDeYtRHcxAmp1J6Q7Hx79Wwem"}
debug   calculator/calls.go:62  consumers trust iterator's init failure {"epoch": 3, "iteration": 2, "error": "no data for 1 iteration in 3 epoch for consumers's trusts"}
debug   calculator/calls.go:62  consumers trust iterator's init failure {"epoch": 3, "iteration": 3, "error": "no data for 2 iteration in 3 epoch for consumers's trusts"}
debug   calculator/calls.go:120 daughter trust iterator's init failure  {"error": "no data in 4 epoch for daughter: o5H2uGUa7cwBvb1GSx62SUL8Mkpp4PbgB1oUstPgnfo8"}
```
Which of these errors are not expected?

**Comment by cthulhu-rider** [comment]
@End-rey 

> debug    network/server.go:1216  p2p notary request      {"error": "already exists in mempool", "hash": "22c5d8035255adf13142eec3e1f1f3256fa8d6874d1806738af44f0601939b09", "main": "dec54cf2feecb63c96ee8ce7f1c372b13bf626e15a0f1acb8dc0991e2811e42d"}

and SN debug messages are np for now

> "error": "test invocation failed: unsupported parameter noderoles.Role"
"error": "test invocation failed: unsupported parameter noderoles.Role"

IR bug, pls report

> can't update list of alphabet nodes in neofs contract  {"error": "could not invoke alphabetUpdate: chain/client: contract execution finished with state FAULT; exception: at instruction 3666 (THROW): unhandled exception: \"this method must be invoked by alphabet\""}

either an IR bug or setup problem. Report too just in case

> can't update list of alphabet nodes in neofs contract  {"error": "could not invoke alphabetUpdate: chain/client: contract execution finished with state FAULT; exception: at instruction 312 (SYSCALL): System.Contract.Call failed: called contract 6a715fe77786ad16a5f543e45c7ba242f196d01b not found: key not found"}

most likely setup problem. Check that NeoFS contract is successfully deployed in the main chain and IR uses correct hash

> can't update inner ring list with new alphabet keys
can't update list of notary nodes in FS chain
can't update list of alphabet nodes in neofs contract

IR should not attempt to do all these in the DevEnv

### Alphabet
u shouldnt merge 2 setups in 1 config [here](https://github.com/nspcc-dev/neofs-dev-env/pull/327/files#diff-c3aa6e6e3e17a7e344fb94dcb15dc7c00f19a23a12990f5cddf9d52ee0a886da), the list must equal `fschain.committee` list foreach setup

### Notary
forgot about this. Add `P2PNotary` list the same as `NeoFSAlphabet`

**Comment by End-rey** [comment]
All the errors above, except for SN debug and `already exists in mempool`, are gone.

**Comment by cthulhu-rider** [comment]
@End-rey dont forget to report bugs https://github.com/nspcc-dev/neofs-node

File: bin/config.sh [hunk 1]
```diff
@@ -4,47 +4,25 @@ echo "Running bin/config.sh"
 
 # Source env settings
 . .env
-. services/ir/.ir.env
 source bin/helper.sh
 
-# NeoGo binary path.
-NEOGO="${NEOGO:-docker exec main_chain neo-go}"
+# NeoADM binary path and config path.
+NEOADM="${NEOADM:-./vendor/neofs-adm}"
```
Comment (carpawell) [inline bin/config.sh:10]: `FS` is important here

File: Makefile [hunk 2]
```diff
@@ -67,6 +69,14 @@ define error_handler
 	fi
 endef
 
+# Check if number of IR nodes from .env file is correct
+.PHONY: check_nodes
+check_nodes:
+	@if ! echo "$(AVAILABLE_NUMBER_OF_NODES)" | grep -wq "$(IR_NUMBER_OF_NODES)"; then \
+		echo "Error: IR_NUMBER_OF_NODES=$(IR_NUMBER_OF_NODES) is not in AVAILABLE_NUMBER_OF_NODES ($(AVAILABLE_NUMBER_OF_NODES))"; \
```
Comment (carpawell) [inline Makefile:76]: can this error line be a little bit simpler? like "invalid IR number X; supported numbers: Y, Z, ..."

File: services/ir1/.hosts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+IPV4_PREFIX.61 ir01.LOCAL_DOMAIN
+IPV4_PREFIX.62 ir02.LOCAL_DOMAIN
+IPV4_PREFIX.63 ir03.LOCAL_DOMAIN
+IPV4_PREFIX.64 ir04.LOCAL_DOMAIN
+IPV4_PREFIX.65 ir05.LOCAL_DOMAIN
+IPV4_PREFIX.66 ir06.LOCAL_DOMAIN
+IPV4_PREFIX.67 ir07.LOCAL_DOMAIN
```
Comment (carpawell) [inline services/ir1/.hosts:7]: why is it placed in `.../ir1/...` dir?

File: services/ir1/artifacts.mk [hunk 4]
```diff
@@ -1,13 +1,13 @@
...
-get.ir: get.cli get.contracts get.adm get.storage prepare.storage
+get.ir1: get.cli get.contracts get.adm get.storage prepare.storage
 
 # Download NeoFS CLI 
 .ONESHELL:
 get.cli: NEOFS_CLI_FILE=./vendor/neofs-cli
 get.cli: NEOFS_CLI_PATH?=
 get.cli:
-	@touch services/ir/.ir.env # https://github.com/docker/compose/issues/3560
+	@touch services/ir${IR_NUMBER_OF_NODES}/.ir.env # https://github.com/docker/compose/issues/3560
```
Comment (carpawell) [inline services/ir1/artifacts.mk:10]: also placed in `.../ir1/...` dir but looks like it should support different numbers too

File: services/ir1/.hosts (hunk 3 above)
Comment (End-rey) [reply services/ir1/.hosts:7]: Should I make a shared folder for such files? For example, just `.../ir/...`.

Comment (carpawell) [reply services/ir1/.hosts]: yes, i would to that, otherwise it confuses

**Comment by cthulhu-rider** [comment]
my trial's summary

### Fetch and run right away
`make up/basic` :ballot_box_with_check: 
`neofs-cli container create` :ballot_box_with_check:
```
$ docker ps
CONTAINER ID   IMAGE                           COMMAND                  CREATED              STATUS                        PORTS     NAMES
c7d39a3c2bba   nspccdev/neofs-storage:0.44.1   "neofs-node --config…"   About a minute ago   Up About a minute (healthy)             s04
ce82d7a1cd85   nspccdev/neofs-storage:0.44.1   "neofs-node --config…"   About a minute ago   Up About a minute (healthy)             s03
410cea04da4b   nspccdev/neofs-storage:0.44.1   "neofs-node --config…"   About a minute ago   Up About a minute (healthy)             s02
087d12987812   nspccdev/neofs-storage:0.44.1   "neofs-node --config…"   About a minute ago   Up About a minute (healthy)             s01
890db6f0c6f7   nspccdev/neofs-ir:0.44.1        "neofs-ir --config /…"   2 minutes ago        Up 2 minutes (healthy)                  ir01
1d8f909711ea   nspccdev/neo-go:0.107.1         "/usr/bin/privnet-en…"   2 minutes ago        Up 2 minutes (healthy)                  main_chain
0c8cf4bc8dd8   debian:10                       "/bin/sleep infinity"    2 minutes ago        Up 2 minutes                            bastion

$ docker volume ls
DRIVER    VOLUME NAME
local     ir1_neofs_chain_ir01
local     ir_neofs_chain_ir01
local     storage_storage_s01
local     storage_storage_s02
local     storage_storage_s03
local     storage_storage_s04
```

### Change num
3:
```
$ make up/basic
Invalid IR number 3; supported numbers: (1 4 7)
make: *** [Makefile:75: check_nodes] Error 1
```
:ballot_box_with_check:

-3:
```
$ make up/basic
Usage: grep [OPTION]... PATTERNS [FILE]...
Try 'grep --help' for more information.
Invalid IR number -3; supported numbers: (1 4 7)
```
:ballot_box_with_check:

4:
```
$ make up/basic
ERROR: for ir01  Cannot create container for service ir01: Conflict. The container name "/ir01" is already in use by container "890db6f0c6f78873a12c51e6e7083d3c34d80e4bbff6d0f4cdfdc7597858bf43". You have to remove (or rename) that container to be able to reuse that name.
Creating ir03 ... done
Creating ir04 ... done
Creating ir02 ... done

ERROR: for ir01  Cannot create container for service ir01: Conflict. The container name "/ir01" is already in use by container "890db6f0c6f78873a12c51e6e7083d3c34d80e4bbff6d0f4cdfdc7597858bf43". You have to remove (or rename) that container to be able to reuse that name.

$ docker ps
CONTAINER ID   IMAGE                           COMMAND                  CREATED          STATUS                     PORTS     NAMES
b81add0ef95f   nspccdev/neofs-ir:0.44.1        "neofs-ir --config /…"   9 minutes ago    Up 9 minutes (unhealthy)             ir03
5f9d1e43ee4c   nspccdev/neofs-ir:0.44.1        "neofs-ir --config /…"   9 minutes ago    Up 9 minutes (unhealthy)             ir02
59dc765ed1bc   nspccdev/neofs-ir:0.44.1        "neofs-ir --config /…"   9 minutes ago    Up 9 minutes (unhealthy)             ir04
c7d39a3c2bba   nspccdev/neofs-storage:0.44.1   "neofs-node --config…"   13 minutes ago   Up 13 minutes (healthy)              s04
ce82d7a1cd85   nspccdev/neofs-storage:0.44.1   "neofs-node --config…"   13 minutes ago   Up 13 minutes (healthy)              s03
410cea04da4b   nspccdev/neofs-storage:0.44.1   "neofs-node --config…"   13 minutes ago   Up 13 minutes (healthy)              s02
087d12987812   nspccdev/neofs-storage:0.44.1   "neofs-node --config…"   13 minutes ago   Up 13 minutes (healthy)              s01
890db6f0c6f7   nspccdev/neofs-ir:0.44.1        "neofs-ir --config /…"   14 minutes ago   Up 14 minutes (healthy)              ir01
0c8cf4bc8dd8   debian:10                       "/bin/sleep infinity"    14 minutes ago   Up 14 minutes                        bastion

$ docker volume ls
DRIVER    VOLUME NAME
local     aio_cache
local     aio_data
local     ir1_neofs_chain_ir01
local     ir4_neofs_chain_ir01
local     ir4_neofs_chain_ir02
local     ir4_neofs_chain_ir03
local     ir4_neofs_chain_ir04
local     ir_neofs_chain_ir01
local     storage_storage_s01
local     storage_storage_s02
local     storage_storage_s03
local     storage_storage_s04
```
3 nodes `unhealthy`, 1 is `healthy`. Container can be created. 3 nodes stuck in with `info	blockchain/blockchain.go:534	waiting for synchronization with the blockchain network...`. I think this is caused by namespace conflict, intersection leads to this behavior. 

devenv keep goin in 1X run mode, this is unexpected to me. On the other side, error was printed. Id prefer purely independent states. Some kind of :stop_sign: to me

###

**Comment by roman-khimov** [comment]
I think this boils down to ability to run independent devenvs of various kinds and requirement to `make down` before switch to some different env. Both are fine for me at this stage and can be solved separately.

**Review by carpawell — APPROVED** [review]
I looks like `services/irX` better be placed inside `services/ir`.

**Pull request merged**